### DISCUSSION TITLE
Is it possible to make a POST request with JSON-type data?

### DISCUSSION THREAD

**Discussion by mithrajuneo** [hahwul/dalfox#556] [Q&A]
I found out that when special characters are included using the -d option, it doesn't behave as expected, despite numerous attempts. For example, using -d '{ "id" : "value", "pass" : "value" }'.

**Comment by adilburaksen** [comment]
It's a bash quoting issue — the shell eats the double quotes before dalfox sees them. Wrap the body in single quotes:

```bash
dalfox "https://target.com/api" \
  -d '{"id":"value","pass":"value"}' \
  -H "Content-Type: application/json"
```

If you want to fuzz a specific field, just drop `FUZZ` where you want it tested:

```bash
dalfox "https://target.com/api" \
  -d '{"id":"FUZZ","pass":"value"}' \
  -H "Content-Type: application/json"
```

dalfox picks up POST automatically when you use `-d`, no need to specify the method separately.

**Comment by TrueNix** [comment]
Adding to @adilburaksen's answer — there are a few more details worth knowing for JSON-body testing:

## Targeting a specific JSON field

Use the `name:location` syntax with `--param` / `-p` and the `json` location to fuzz a specific field while keeping the rest of the body static:

```bash
dalfox "https://target.com/api"   -d '{"id":"value","pass":"value"}'   -H "Content-Type: application/json"   -p "pass:json"
```

This fuzzes only the `pass` field's value, leaving `id` untouched. Useful when the endpoint has required fields that must be valid for the request to reach the vulnerable code path.

## Multiple JSON fields

```bash
dalfox "https://target.com/api"   -d '{"id":"value","pass":"value","name":"value"}'   -H "Content-Type: application/json"   -p "pass:json" -p "name:json"
```

## Nested JSON

For nested objects, Dalfox handles the top-level keys. If your JSON is `{"user":{"name":"value"}}`, you can target `name` in the `json` location:

```bash
dalfox "https://target.com/api"   -d '{"user":{"name":"FUZZ"}}'   -H "Content-Type: application/json"
```

Put `FUZZ` as a literal placeholder in the body where you want the payload injected — Dalfox replaces it during fuzzing.

## Method override

`-d` implies POST, but you can override with `-X` if the endpoint expects PUT or PATCH:

```bash
dalfox "https://target.com/api"   -d '{"id":"FUZZ"}'   -H "Content-Type: application/json"   -X PUT
```

## Quick note on bash quoting

The original issue was a shell quoting problem, not a Dalfox limitation. Single quotes preserve the JSON literally; double quotes let bash interpret `$"\` etc. If your JSON contains single quotes (rare but possible in string values), use a heredoc or escape carefully:

```bash
dalfox "https://target.com/api"   -d '{"name":"O\"Brien"}'   -H "Content-Type: application/json"
```

Or write the body to a file and pipe it in if the quoting gets hairy.
